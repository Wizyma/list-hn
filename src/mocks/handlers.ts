import { rest } from "msw";

import { BASE_URL } from "$src/constants";
import { StoryResponse, TopStoriesResponse } from "$src/models/api";

export const handlers = [
  rest.get<TopStoriesResponse>(
    `${BASE_URL}/topstories.json?print=pretty`,
    (req, res, ctx) => {
      return res(
        // Respond with a 200 status code
        ctx.status(200),
        ctx.json([
          33786474, 33787218, 33785631, 33786773, 33783504, 33785904, 33784549,
          33773135, 33785614, 33775052, 33782010, 33785870, 33774770, 33782585,
          33782946, 33763127, 33784271, 33762239, 33774514, 33784432,
        ])
      );
    }
  ),

  rest.get<StoryResponse, { id: string }>(
    `${BASE_URL}/item/:id`,
    (req, res, ctx) => {
      const { id } = req.params;
      const numberId = parseInt(id.split(".")[0]);

      if (numberId) {
        return res(
          ctx.status(200),
          ctx.json({
            by: "dhouston",
            descendants: 71,
            id: numberId,
            kids: [9224],
            score: 104,
            time: 1175714200,
            title: "My YC app: Dropbox - Throw away your USB drive",
            type: "story",
            url: "http://www.getdropbox.com/u/2/screencast.html",
          })
        );
      }
    }
  ),
];
