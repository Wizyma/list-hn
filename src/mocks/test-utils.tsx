import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import React, { FC } from "react";

export const testingClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      cacheTime: Infinity,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error:
      process.env.APP_ENV === "test"
        ? () => {
            return null;
          }
        : console.error,
  },
});

export const AllTheProviders = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  return (
    <QueryClientProvider client={testingClient}>{children}</QueryClientProvider>
  );
};

export * from "@testing-library/react";
