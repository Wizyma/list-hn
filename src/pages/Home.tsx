import { useCallback, useEffect, useState } from "react";
import ReactPaginate from "react-paginate";

import "./home.css";
import { Feeds } from "$components/Feeds";
import { useBatchStories } from "$hooks/useBatchStories";
import { useTopStories } from "$hooks/useTopStories";
import { Cursor } from "$models/cursor";

function Home() {
  const [cursor, setCursor] = useState<Cursor<number> | null>(null);
  const { data, status } = useTopStories();
  const { status: batchStoriesStatus } = useBatchStories(cursor?.data);

  const onPageChange = useCallback(
    ({ selected }: { selected: number }) => {
      if (data) {
        setCursor(data[selected]);
      }
    },
    [data]
  );

  useEffect(() => {
    if (!cursor && data?.length) {
      setCursor(data[0]);
    }
  }, [data]);

  return (
    <div className="App">
      <h1 className="title">HackerNews List</h1>
      <Feeds
        currentCursor={cursor?.currentCursor ?? 0}
        isLoading={status === "fetching" || batchStoriesStatus === "fetching"}
        ids={cursor?.data as number[]}
      />
      <ReactPaginate
        breakLabel="..."
        nextLabel=">"
        onPageChange={onPageChange}
        pageRangeDisplayed={5}
        pageCount={data?.length ?? 0}
        previousLabel="<"
      />
    </div>
  );
}

export default Home;
