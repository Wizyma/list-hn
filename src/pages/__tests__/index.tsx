import Home from "../Home";

import { server } from "$src/mocks/server";
import {
  AllTheProviders,
  render,
  waitForElementToBeRemoved,
} from "$src/mocks/test-utils";

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe(Home.name, () => {
  it("should render in loading state", () => {
    const { asFragment, getAllByTestId } = render(<Home />, {
      wrapper: ({ children }) => <AllTheProviders>{children}</AllTheProviders>,
    });

    expect(asFragment()).toMatchSnapshot();
    expect(getAllByTestId("skeleton").length).toBe(10);
  });

  it("should render the list elements", async () => {
    const { asFragment, getAllByText, getAllByTestId } = render(<Home />, {
      wrapper: ({ children }) => <AllTheProviders>{children}</AllTheProviders>,
    });

    await waitForElementToBeRemoved(getAllByTestId("skeleton"));

    expect(asFragment()).toMatchSnapshot();
    expect(getAllByText("104 points").length).toBe(10);
  });
});
