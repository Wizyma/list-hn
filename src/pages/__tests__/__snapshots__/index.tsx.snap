// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render in loading state 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1
      class="title"
    >
      HackerNews List
    </h1>
    <div>
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
      <span
        data-testid="skeleton"
      />
    </div>
    <ul
      aria-label="Pagination"
      role="navigation"
    >
      <li
        class="previous disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          class=" "
          rel="prev"
          role="button"
          tabindex="-1"
        >
          &lt;
        </a>
      </li>
      <li
        class="next"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          class=""
          rel="next"
          role="button"
          tabindex="0"
        >
          &gt;
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Home should render the list elements 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1
      class="title"
    >
      HackerNews List
    </h1>
    <div
      class="feeds"
    >
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            1
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            2
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            3
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            4
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            5
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            6
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            7
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            8
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            9
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
      <a
        href="http://www.getdropbox.com/u/2/screencast.html"
        target="_blank"
      >
        <div>
          <span>
            10
          </span>
          <p>
            My YC app: Dropbox - Throw away your USB drive
          </p>
          <span>
            (www.getdropbox.com)
          </span>
        </div>
        <div>
          <span>
            104 points
          </span>
          <span>
            |
          </span>
          <span>
            by dhouston
          </span>
          <span>
            |
          </span>
          <span />
          <span>
            |
          </span>
          <span>
            71 comments
          </span>
        </div>
      </a>
    </div>
    <ul
      aria-label="Pagination"
      role="navigation"
    >
      <li
        class="previous disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          class=" "
          rel="prev"
          role="button"
          tabindex="-1"
        >
          &lt;
        </a>
      </li>
      <li
        class="selected"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          rel="canonical"
          role="button"
          tabindex="-1"
        >
          1
        </a>
      </li>
      <li>
        <a
          aria-label="Page 2"
          rel="next"
          role="button"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li
        class="next"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          class=""
          rel="next"
          role="button"
          tabindex="0"
        >
          &gt;
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
